lib/sprockets/utils.rb

```diff
@@ -97,11 +100,21 @@ def string_end_with_semicolon?(str)
     #
     # Returns buf String.
     def concat_javascript_sources(buf, source)
-      if buf.bytesize > 0
-        buf << ";" unless string_end_with_semicolon?(buf)
-        buf << "\n" unless buf.end_with?("\n")
+      if source.bytesize > 0
+        buf << source
+
+        # If the source contains non-ASCII characters, indexing on it becomes O(N).
+        # This will lead to O(N^2) performance in string_end_with_semicolon?, so we should use 32 bit encoding to make sure indexing stays O(1)
+        source = source.encode(Encoding::UTF_32LE) unless source.ascii_only?
```

Comment:
I mean, before this patch the source was append after the `;` and `\n`, now it is before. Can we improve the performance without changing the order of the behavior?